File: src/main/java/ffblockly/meta/AvOption.java
package ffblockly.meta;

import java.util.List;
import java.util.regex.Pattern;

import static ffblockly.meta.AvOption.AvOptionField.*;

/**
 * Created by jim on 9/12/2016.
 */
public interface AvOption {

    Pattern AV_OPT_PAT = Pattern.compile("^\\s+(?<" +
            name +
            ">[\\w+-]+)(\\s+\\<(?<" +
            type +
            ">\\w+)\\>)?\\s+(?<" +
            targets +
            ">[\\w.]{8})(\\s+(?<" +
            description +
            ">[^(]+)(\\s+\\(from\\s+(?<" +
            range1 +
            ">.*)\\s+to\\s+(?<" +
            range2 +
            ">[^)]+)\\))?(\\s+\\(default\\s+(?<" +
            def +
            ">[^)]+)\\))?)?$");

    String getName();

    String getType();

    List<AvTarget> getTargets();

    String getDescription();

    String getRange1();

    String getRange2();

    String getDef();

    List<AvOption> getChildren();

    Boolean getEncoding();

    Boolean getDecoding();

    Boolean getFiltering();

    Boolean getVideo();

    Boolean getAudio();

    Boolean getSubtitle();

    Boolean getExport();

    Boolean getReadonly();

    void addChild(AvOption o);

    enum AvTarget {
        Encoding, Decoding, Filtering, Video, Audio, Subtitle, Export, Readonly
    }

    enum AvOptionField {
        name, type, targets, description, range1, range2, def
    }
}

File: src/main/java/ffblockly/meta/AvOption.kt
package ffblockly.meta

import java.util.regex.Pattern

/**
 * Created by jim on 9/12/2016.
 */
interface AvOption {
    val optionName: String?
    val type: String?
    val targets: List<AvTarget>?
    val description: String?
    val range1: String?
    val range2: String?
    val def: String?
    val children: List<AvOption>?
    val encoding: Boolean?
    val decoding: Boolean?
    val filtering: Boolean?
    val video: Boolean?
    val audio: Boolean?
    val subtitle: Boolean?
    val export: Boolean?
    val readonly: Boolean?
    fun addChild(o: AvOption)
    enum class AvTarget {
        Encoding, Decoding, Filtering, Video, Audio, Subtitle, Export, Readonly
    }

    enum class AvOptionField {
        optionName, type, targets, description, range1, range2, def
    }

    companion object {
        val AV_OPT_PAT = Pattern.compile("^\\s+(?<" + AvOptionField.optionName + ">[\\w+-]+)(\\s+\\<(?<" + AvOptionField.type + ">\\w+)\\>)?\\s+(?<" + AvOptionField.targets + ">[\\w.]{8})(\\s+(?<" + AvOptionField.description + ">[^(]+)(\\s+\\(from\\s+(?<" + AvOptionField.range1 + ">.*)\\s+to\\s+(?<" + AvOptionField.range2 + ">[^)]+)\\))?(\\s+\\(default\\s+(?<" + AvOptionField.def + ">[^)]+)\\))?)?$")
    }
}